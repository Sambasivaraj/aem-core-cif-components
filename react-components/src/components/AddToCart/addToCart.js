/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 ~ Copyright 2021 Adobe
 ~
 ~ Licensed under the Apache License, Version 2.0 (the "License");
 ~ you may not use this file except in compliance with the License.
 ~ You may obtain a copy of the License at
 ~
 ~     http://www.apache.org/licenses/LICENSE-2.0
 ~
 ~ Unless required by applicable law or agreed to in writing, software
 ~ distributed under the License is distributed on an "AS IS" BASIS,
 ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 ~ See the License for the specific language governing permissions and
 ~ limitations under the License.
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/

import React, { useCallback, useEffect, useRef, useState } from 'react';
import { useIntl } from 'react-intl';
import { useMutation } from '@apollo/client';
import { useCartContext } from '@magento/peregrine/lib/context/cart';
import { productMapper, bundledProductMapper } from '../Minicart/useMinicart';
import MUTATION_ADD_TO_CART from '../../queries/mutation_add_to_cart.graphql';
import MUTATION_ADD_BUNDLE_TO_CART from '../../queries/mutation_add_bundle_to_cart.graphql';
import MUTATION_ADD_VIRTUAL_TO_CART from '../../queries/mutation_add_virtual_to_cart.graphql';
import MUTATION_ADD_SIMPLE_AND_VIRTUAL_TO_CART from '../../queries/mutation_add_simple_and_virtual_to_cart.graphql';

const PRODUCT_DATA_UPDATE = 'aem.cif.internal.add-to-cart.state-changed';

const useEventListener = props => {
    const { eventName, eventListener } = props;
    const ref = useRef(null);
    const callback = useCallback(node => {
        // remove listener from previous ref, if any
        if (ref.current) {
            ref.current.removeEventListener(eventName, eventListener);
        }
        // add listner to new ref, if any
        if (node) {
            node.addEventListener(eventName, eventListener);
        }
        // setRef
        ref.current = node;
    });
    return [callback];
};

const AddToCart = props => {
    const intl = useIntl();
    const [items, setItems] = useState([]);
    const [disabled, setDisabled] = useState(true);
    const handleProductDataUpdate = event => setItems(event.detail);
    const [buttonRef] = useEventListener({ eventName: PRODUCT_DATA_UPDATE, eventListener: handleProductDataUpdate });
    const [{ cartId }] = useCartContext();
    const [addToCartMutation] = useMutation(MUTATION_ADD_TO_CART);
    const [addBundleItemMutation] = useMutation(MUTATION_ADD_BUNDLE_TO_CART);
    const [addVirtualItemMutation] = useMutation(MUTATION_ADD_VIRTUAL_TO_CART);
    const [addSimpleAndVirtualItemMutation] = useMutation(MUTATION_ADD_SIMPLE_AND_VIRTUAL_TO_CART);

    const buttonContent = props.children ? (
        props.children
    ) : (
        <span>{intl.formatMessage({ id: 'add-to-cart:label', defaultMessage: 'Add to Cart' })}</span>
    );

    const handleAddToCart = useCallback(async () => {
        const physicalCartItems = items.filter(item => !item.virtual).map(productMapper);
        const virtualCartItems = items.filter(item => item.virtual).map(productMapper);
        const bundleCartItems = items.filter(item => item.bundle).map(bundledProductMapper);

        if (bundleCartItems.length > 0) {
            await addBundleItemMutation({
                variables: {
                    cartId,
                    cartItems: bundleCartItems
                }
            });
        } else if (virtualCartItems.length > 0 && physicalCartItems.length > 0) {
            await addSimpleAndVirtualItemMutation({
                variables: {
                    cartId,
                    virtualCartItems: virtualCartItems,
                    simpleCartItems: physicalCartItems
                }
            });
        } else if (virtualCartItems.length > 0) {
            await addVirtualItemMutation({
                variables: {
                    cartId,
                    cartItems: virtualCartItems
                }
            });
        } else if (physicalCartItems.length > 0) {
            await addToCartMutation({
                variables: {
                    cartId,
                    cartItems: physicalCartItems
                }
            });
        }

        if (props.onAddToCart) {
            props.onAddToCart(items);
        } else if (items.length > 0) {
            // dispatch the event that other wise the product detail add to cart function would have
            // fired
            document.dispatchEvent(
                new CustomEvent('aem.cif.add-to-cart', {
                    detail: items
                })
            );
        }
    }, [props.onAddToCart, items, cartId]);

    useEffect(() => {
        const newItems = typeof props.items === 'string' ? JSON.parse(props.items) : props.items;
        setItems(newItems);
    }, [props.items]);

    useEffect(() => {
        setDisabled(props.disabled || items.length === 0);
    }, [props.disabled, items]);

    return (
        <button
            ref={buttonRef}
            className="button__root_highPriority button__root clickable__root button__filled"
            type="button"
            onClick={handleAddToCart}
            disabled={disabled}>
            <span className="button__content">{buttonContent}</span>
        </button>
    );
};

export default AddToCart;
