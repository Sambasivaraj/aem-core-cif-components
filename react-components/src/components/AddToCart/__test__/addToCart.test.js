/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 ~ Copyright 2021 Adobe
 ~
 ~ Licensed under the Apache License, Version 2.0 (the "License");
 ~ you may not use this file except in compliance with the License.
 ~ You may obtain a copy of the License at
 ~
 ~     http://www.apache.org/licenses/LICENSE-2.0
 ~
 ~ Unless required by applicable law or agreed to in writing, software
 ~ distributed under the License is distributed on an "AS IS" BASIS,
 ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 ~ See the License for the specific language governing permissions and
 ~ limitations under the License.
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
jest.mock('@apollo/client', () => ({
    ...jest.requireActual('@apollo/client'),
    useMutation: jest.fn()
}));

import React from 'react';
import { render, fireEvent, wait } from 'test-utils';

import AddToCart from '../addToCart';

import { useMutation } from '@apollo/client';

const simpleItem = {
    sku: 'simple-sku',
    quantity: 1
};

const bundleItem = {
    sku: 'bundle-sku',
    quantity: 2,
    bundle: true,
    options: [
        {
            foo: 'bar'
        }
    ]
};

const virtualItem = {
    sku: 'virtual-sku',
    quantity: 3,
    virtual: true
};

const dispatchEventSpy = jest.spyOn(document, 'dispatchEvent');

const addToCartMutationMock = jest.fn();
const addBundleToCartMutationMock = jest.fn();
const addVirtualToCartMutationMock = jest.fn();
const addSimpleAndVirtualToCartMutationMock = jest.fn();

jest.mock('../../../queries/mutation_add_to_cart.graphql', () => 'addToCart');
jest.mock('../../../queries/mutation_add_bundle_to_cart.graphql', () => 'addBundleToCart');
jest.mock('../../../queries/mutation_add_virtual_to_cart.graphql', () => 'addVirtualToCart');
jest.mock('../../../queries/mutation_add_simple_and_virtual_to_cart.graphql', () => 'addSimpleAndVirtualToCart');

useMutation.mockImplementation(mutation => {
    const { useMutation: actualUseMutation } = jest.requireActual('@apollo/client');
    if (mutation === 'addToCart') {
        return [addToCartMutationMock];
    } else if (mutation === 'addBundleToCart') {
        return [addBundleToCartMutationMock];
    } else if (mutation === 'addVirtualToCart') {
        return [addVirtualToCartMutationMock];
    } else if (mutation === 'addSimpleAndVirtualToCart') {
        return [addSimpleAndVirtualToCartMutationMock];
    } else {
        return actualUseMutation(mutation);
    }
});

describe('<AddToCart>', () => {
    afterEach(() => {
        dispatchEventSpy.mockClear();
        addToCartMutationMock.mockClear();
        addBundleToCartMutationMock.mockClear();
        addVirtualToCartMutationMock.mockClear();
        addSimpleAndVirtualToCartMutationMock.mockClear();
    });

    it('renders the component for empty items', () => {
        const { asFragment, getByRole } = render(<AddToCart items={[]} />);

        expect(asFragment()).toMatchSnapshot();
        expect(getByRole('button')).toBeDisabled();
    });

    it('renders the component with disabled property', () => {
        const { asFragment, getByRole } = render(<AddToCart items={[simpleItem]} disabled />);

        expect(asFragment()).toMatchSnapshot();
        expect(getByRole('button')).toBeDisabled();
    });

    it('renders the component with simple item', () => {
        const { asFragment, getByRole } = render(<AddToCart items={[simpleItem]} />);

        expect(asFragment()).toMatchSnapshot();
        expect(getByRole('button')).not.toBeDisabled();
    });

    it('add a simple product to cart', async () => {
        const { getByRole } = render(<AddToCart items={[simpleItem]} />);

        fireEvent.click(getByRole('button'));
        expect(addToCartMutationMock).toHaveBeenCalledTimes(1);
        expect(addToCartMutationMock).toHaveBeenCalledWith({
            variables: {
                cartId: null,
                cartItems: [
                    {
                        data: {
                            sku: simpleItem.sku,
                            quantity: parseFloat(simpleItem.quantity)
                        }
                    }
                ]
            }
        });
        await wait(() => {
            expect(dispatchEventSpy).toHaveBeenCalledTimes(1);
            expect(dispatchEventSpy).toHaveBeenCalledWith(new CustomEvent('aem.cif.after-add-to-cart'));
        });
    });

    it('add a bundle product to cart', async () => {
        const { getByRole } = render(<AddToCart items={[bundleItem]} />);

        fireEvent.click(getByRole('button'));
        expect(addBundleToCartMutationMock).toHaveBeenCalledTimes(1);
        expect(addBundleToCartMutationMock).toHaveBeenCalledWith({
            variables: {
                cartId: null,
                cartItems: [
                    {
                        data: {
                            sku: bundleItem.sku,
                            quantity: parseFloat(bundleItem.quantity)
                        },
                        bundle_options: bundleItem.options
                    }
                ]
            }
        });
        await wait(() => {
            expect(dispatchEventSpy).toHaveBeenCalledTimes(1);
            expect(dispatchEventSpy).toHaveBeenCalledWith(new CustomEvent('aem.cif.after-add-to-cart'));
        });
    });

    it('add a virtual product to cart', async () => {
        const { getByRole } = render(<AddToCart items={[virtualItem]} />);

        fireEvent.click(getByRole('button'));
        expect(addVirtualToCartMutationMock).toHaveBeenCalledTimes(1);
        expect(addVirtualToCartMutationMock).toHaveBeenCalledWith({
            variables: {
                cartId: null,
                cartItems: [
                    {
                        data: {
                            sku: virtualItem.sku,
                            quantity: parseFloat(virtualItem.quantity)
                        }
                    }
                ]
            }
        });
        await wait(() => {
            expect(dispatchEventSpy).toHaveBeenCalledTimes(1);
            expect(dispatchEventSpy).toHaveBeenCalledWith(new CustomEvent('aem.cif.after-add-to-cart'));
        });
    });

    it('add a simple product and a virtual product to cart', async () => {
        const { getByRole } = render(<AddToCart items={[simpleItem, virtualItem]} />);

        fireEvent.click(getByRole('button'));
        expect(addSimpleAndVirtualToCartMutationMock).toHaveBeenCalledTimes(1);
        expect(addSimpleAndVirtualToCartMutationMock).toHaveBeenCalledWith({
            variables: {
                cartId: null,
                virtualCartItems: [
                    {
                        data: {
                            sku: virtualItem.sku,
                            quantity: parseFloat(virtualItem.quantity)
                        }
                    }
                ],
                simpleCartItems: [
                    {
                        data: {
                            sku: simpleItem.sku,
                            quantity: parseFloat(simpleItem.quantity)
                        }
                    }
                ]
            }
        });
        await wait(() => {
            expect(dispatchEventSpy).toHaveBeenCalledTimes(1);
            expect(dispatchEventSpy).toHaveBeenCalledWith(new CustomEvent('aem.cif.after-add-to-cart'));
        });
    });

    it('add a simple product and provides callback', async () => {
        const onAddToCartMock = jest.fn();
        const { getByRole } = render(<AddToCart items={[simpleItem]} onAddToCart={onAddToCartMock} />);

        fireEvent.click(getByRole('button'));
        await wait(() => {
            expect(dispatchEventSpy).toHaveBeenCalledTimes(0);
            expect(onAddToCartMock).toHaveBeenCalledTimes(1);
            expect(onAddToCartMock).toHaveBeenCalledWith([simpleItem]);
        });
    });
});
