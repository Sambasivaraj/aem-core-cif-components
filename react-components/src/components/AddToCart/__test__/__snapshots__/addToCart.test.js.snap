// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddToCart> renders the component for empty items 1`] = `
<DocumentFragment>
  <button
    class="button__root_highPriority button__root clickable__root button__filled"
    disabled=""
    type="button"
  >
    <span
      class="button__content"
    >
      <span>
        Add to Cart
      </span>
    </span>
  </button>
</DocumentFragment>
`;

exports[`<AddToCart> renders the component with disabled property 1`] = `
<DocumentFragment>
  <button
    class="button__root_highPriority button__root clickable__root button__filled"
    disabled=""
    type="button"
  >
    <span
      class="button__content"
    >
      <span>
        Add to Cart
      </span>
    </span>
  </button>
</DocumentFragment>
`;

exports[`<AddToCart> renders the component with simple item 1`] = `
<DocumentFragment>
  <button
    class="button__root_highPriority button__root clickable__root button__filled"
    type="button"
  >
    <span
      class="button__content"
    >
      <span>
        Add to Cart
      </span>
    </span>
  </button>
</DocumentFragment>
`;
